builtin/mainmenu/content/dlg_contentdb.lua

```diff
@@ -176,72 +128,151 @@ local function load()
 	end
 	assert(contentdb.load_ok)
 
 	contentdb.update_paths()
 
+	local num_per_page = dlgdata.num_per_page
 	dlgdata.pagemax = math.max(math.ceil(#contentdb.packages / num_per_page), 1)
 	if cur_page > dlgdata.pagemax then
 		cur_page = 1
 	end
 
-	local W = 15.75
-	local H = 9.5
-	local formspec = {
-		"formspec_version[6]",
-		"size[15.75,9.5]",
-		core.settings:get_bool("touch_gui") and "padding[0.01,0.01]" or "position[0.5,0.55]",
+	local W = size.x - window_padding.x * 2
+	local H = size.y - window_padding.y * 2
+
+	local category_x = 0
+	local number_category_buttons = 4
+	local max_button_w = (W - 0.375 - 0.25 - 7) / number_category_buttons
+	local category_button_w = math.min(max_button_w, 3)
+	local function make_category_button(name, label, selected)
+		category_x = category_x + 1
+		local color = selected and mt_color_green or ""
+		return ("style[%s;bgcolor=%s]button[%f,0;%f,0.8;%s;%s]"):format(name, color,
+				(category_x - 1) * category_button_w, category_button_w, name, label)
+	end
 
-		"style[status,downloading,queued;border=false]",
 
-		"container[0.375,0.375]",
-		"field[0,0;7.225,0.8;search_string;;", core.formspec_escape(search_string), "]",
+	local selected_type = filter_type
+
+	local search_box_width = W - 0.375 - 0.25 - 2*0.8
+			- number_category_buttons * category_button_w
+	local formspec = {
+		"formspec_version[7]",
+		"size[", size.x, ",", size.y, "]",
+		"padding[0,0]",
+		"bgcolor[;true]",
+
+		"container[", window_padding.x, ",", window_padding.y, "]",
+
+		-- Top-left: categories
+		make_category_button("type_all", fgettext("All"), selected_type == nil),
+		make_category_button("type_game", fgettext("Games"), selected_type == "game"),
+		make_category_button("type_mod", fgettext("Mods"), selected_type == "mod"),
+		make_category_button("type_txp", fgettext("Texture Packs"), selected_type == "txp"),
+
+		-- Top-right: Search
+		"container[", W - search_box_width - 0.8*2, ",0]",
+		"field[0,0;", search_box_width, ",0.8;search_string;;", core.formspec_escape(search_string), "]",
 		"field_enter_after_edit[search_string;true]",
-		"image_button[7.3,0;0.8,0.8;", core.formspec_escape(defaulttexturedir .. "search.png"), ";search;]",
-		"image_button[8.125,0;0.8,0.8;", core.formspec_escape(defaulttexturedir .. "clear.png"), ";clear;]",
-		"dropdown[9.175,0;2.7875,0.8;type;", table.concat(filter_types_titles, ","), ";", filter_type, "]",
+		"image_button[", search_box_width, ",0;0.8,0.8;",
+			core.formspec_escape(defaulttexturedir .. "search.png"), ";search;]",
+		"image_button[", search_box_width + 0.8, ",0;0.8,0.8;",
+			core.formspec_escape(defaulttexturedir .. "clear.png"), ";clear;]",
 		"container_end[]",
 
-		-- Page nav buttons
-		"container[0,", H - 0.8 - 0.375, "]",
-		"button[0.375,0;5,0.8;back;", fgettext("Back to Main Menu"), "]",
+		-- Bottom strip start
+		"container[0,", H - 0.8, "]",
+		"button[0,0;4,0.8;back;", fgettext("Back to Main Menu"), "]",
```

Comment:
all the other back buttons are just labelled "Back". could do this here too for consistency.